### DISCUSSION TITLE
MappedAsDataclass TypeError when use sort_order in SQLAlchemy2.0.4

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#9333] [Usage Questions] [answered]
Mixin classes:
```python
id_key = Annotated[int, mapped_column(primary_key=True, index=True, autoincrement=True)]


class _BaseMixin:

    create_user: Mapped[int] = mapped_column(sort_order=9999)
    update_user: Mapped[Optional[int]] = mapped_column(default=None, sort_order=9999)
    created_time: Mapped[datetime] = mapped_column(init=False, sort_order=9999, default=func.now())
    updated_time: Mapped[Optional[datetime]] = mapped_column(init=False, sort_order=9999, onupdate=func.now())


class MappedBase(DeclarativeBase):

    id: Mapped[id_key] = mapped_column(sort_order=-9999)

    @declared_attr.directive
    def __tablename__(cls) -> str:  # noqa
        return cls.__name__.lower()


class DataClassBase(MappedAsDataclass, MappedBase):

    id: Mapped[id_key] = mapped_column(init=False, sort_order=-9999, use_existing_column=False)

    __abstract__ = True


class Base(DataClassBase, _BaseMixin):

    __abstract__ = True
```
This is a sort_order test. E.g.:
```python
class User(Base):
    __tablename__ = 'sys_user'
    
    uid: Mapped[str] = mapped_column(String(50), init=False, default_factory=uuid4, unique=True)
    username: Mapped[str] = mapped_column(String(20), unique=True, index=True)
    password: Mapped[str] = mapped_column(String(255))
    email: Mapped[str] = mapped_column(String(50), unique=True, index=True)
    time_joined: Mapped[datetime] = mapped_column(init=False, default=func.now())
    last_login: Mapped[Optional[datetime]] = mapped_column(init=False, onupdate=func.now())
```
When use alembic create tables or `print(User.__tables__.columns)`: `TypeError: non-default argument 'username' follows default argument`. But in `id` and `uid` columns, used `init=False`, So they will not be included as default value option columns, and all columns of `_BaseMixin` will be listed last, is this error an internal handling exception? Convert to dataclasses before sorting?

**Comment by CaselIT** [comment]
Hi,

This is expected, and is dictated by how dataclass work. The same code would error if you tried to create plain dataclasses.  
Personally I don't think sqlalchemy should diverge from the standard dataclass behaviour.


The sort_order option that was added in 2.0.4 is only used control how table columns will be generated in the db.

To fix the error you would do exactly like you would with plain dataclasses, that is add kw_only=True in the fields in the mixin

**Comment by wu-clan** [comment]
It doesn't seem to work in python 3.8, this parameter requires python 3.10+

**Reply by CaselIT** [reply]
I think you can either set init=False in the class that have this problrm, update python version or not use dataclasses

**Reply by wu-clan** [reply]
This is not what I expected

**Reply by CaselIT** [reply]
I'm not sure what you expect. This is how dataclass work. Here is your example using plain dataclasses. It has the same exact error
```py
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional
from uuid import uuid4

@dataclass
class _BaseMixin:
    create_user: int
    update_user: Optional[int] = field(default=None)
    created_time: datetime = field(init=False, default="func.now()")
    updated_time: Optional[datetime] = field(init=False)

@dataclass
class MappedBase:
    id: int

@dataclass
class DataClassBase(MappedBase):
    id:int = field(init=False)

@dataclass
class Base(DataClassBase, _BaseMixin):
    pass

@dataclass
class User(Base):
    uid: str = field(init=False, default_factory=uuid4)
    username: str
    password: str
    email: str
    time_joined: datetime = field(init=False, default="func.now()")
    last_login: Optional[datetime] = field(init=False)
```

**Comment by wu-clan** [comment]
Hi, 
without using `sort_order` parameter, I tested the inheritance order of columns in `SQLAlchemy2.0.b1`, for the columns of subclasses, they are always ordered after the columns of the parent class, but from `SQLAlchemy2.0.0` onwards, the columns of subclasses columns start to be listed first, which is exactly the problem when I upgraded to `2.0.4`, because when i use `2.0.b1`, in order for the `id` primary key to be listed first, I forced the user to add it manually. Will this be converted to issue processing?

**Reply by CaselIT** [reply]
That was an intentional change. You can use sort_order to control how the columns are ordered in the table.

**Reply by wu-clan** [reply]
Well, I will try the existing logic to modify my code

**Reply by CaselIT** [reply]
I'm not sure what you are trying to do.
Is the problem the order of the columns when creating the table in the db? If so you can use sort_order to control it.
Or the fact that dataclass does not accept fields without default after default ones? In this case it really depends on how dataclass works, so you will probably need to work around it in some way like you would using plain dataclasses

**Reply by wu-clan** [reply]
Sorry, my description is not clear enough. I string the context we discussed. In fact, what I want to say is ’sqlalchemy2.0.0‘, the inheritance order of the subclass column has changed. At this time, there is no ’sort_order‘. The subclass column begins to automatically be in front of the parent class column, and when b1, the subclass column is automatically located after the parent column.

Previously, I relied on the inheritance order of b1 to design the data model class, but due to the change of 2.0.0, some of my columns with default fields changed from the back to the front, which led to a data class verification error, so I think the change in the inheritance order brought about by 2.0.0 is wrong, but you mentioned that this is an intentional change, but ’sort_order‘ is only sorting at the database level.It has nothing to do with the original column order of the class itself. 

Even if 2.0.0 does not change the order of inheritance columns, it will not affect the use of the ‘sort_order’ parameter added to 2.0.4. Another view is that the inheritance order pattern in b1 itself is wrong and does not conform to the original working mode of the ordinary data class, so 2.0.0 fixed it, and I designed the model on the basis of errors. I am not following the 2.0.0 changelog, which may have explained this error point.

**Reply by CaselIT** [reply]
> Previously, I relied on the inheritance order of b1 to design the data model class, but due to the change of 2.0.0, some of my columns with default fields changed from the back to the front, which led to a data class verification error, so I think the change in the inheritance order brought about by 2.0.0 is wrong, but you mentioned that this is an intentional change, but ’sort_order‘ is only sorting at the database level.

I have to verify this, but I don't think the order of the fields in a dataclass ever changed, only the order of the columns in the table that's generated.

Note that there is no such ordering limitation when not using dataclasses, so maybe the behavior you are seeing is because in b1 normal classes were used (normal meaning not dataclasses)

I'll try verifying the behaviour

**Reply by CaselIT** [reply]
Ok I understand what's happening now:
in b1 the generation of a dataclass was quite different compared to how it's is now, in particular sqlalchemy was re-assigning all attributes to the class, with the order that did not follow a very consistent behavior. That was changed to more closely resemble how dataclass worked.

I think the current behaviour is "better" in the sense that does what dataclass does and is not sqlalchemy incantation of it, but maybe a flag could be provided to indicate if dataclasses should respect the order obtained by sort_order. It would probably break code completition tools/type checkers though.
@zzzeek What do you think mike?

@wu-clan note that since 2.0.4 added support for custom dataclass callable, you can also re-order the attributes as you prefer.
For example this order as last all the keys of _BaseMixin:

```py
def my_dataclass(cls, **kw):
    if cls.__name__ == "User":
        o = ["create_user", "update_user", "created_time", "updated_time"]
        new_anno = {
            k: v for (k, v) in cls.__annotations__.items() if k not in o
        }
        for k in cls.__annotations__:
            if k not in new_anno:
                new_anno[k] = cls.__annotations__[k]
        cls.__annotations__ = new_anno
    return dataclasses.dataclass(cls, **kw)

class DataClassBase(MappedAsDataclass, MappedBase, dataclass_callable=my_dataclass):

    id: Mapped[id_key] = mapped_column(init=False, sort_order=-9999, use_existing_column=False)

    __abstract__ = True
```

do note that code completition will probably suggest args in init in the wrong order and / or type checkers may report errors

**Reply by zzzeek** [reply]
I think the both the old behavior and the new beahvior are both bugs and all attributes on _BaseMixin and MappedBase should be ignored as far as dataclasses are concerned and not be part of the constructor, since neither of them extend from MappedAsDataclass.    Pyright rightly shows the constructor of User having only username, password, email. the pure dataclasses test should look like this:

```py
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional
from uuid import uuid4

class _BaseMixin:
    create_user: int
    update_user: Optional[int] = field(default=None)
    created_time: datetime = field(init=False, default="func.now()")
    updated_time: Optional[datetime] = field(init=False)

class MappedBase:
    id: int

@dataclass
class DataClassBase(MappedBase):
    id:int = field(init=False)

@dataclass
class Base(DataClassBase, _BaseMixin):
    pass

@dataclass
class User(Base):
    uid: str = field(init=False, default_factory=uuid4)
    username: str
    password: str
    email: str
    time_joined: datetime = field(init=False, default="func.now()")
    last_login: Optional[datetime] = field(init=False)


import inspect

print(inspect.getfullargspec(User.__init__))
```

argspec is:

```
FullArgSpec(args=['self', 'username', 'password', 'email'], varargs=None, varkw=None, defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={'return': None, 'username': <class 'str'>, 'password': <class 'str'>, 'email': <class 'str'>})

```

so no, there'd be no TypeError, but also, create_user, id, update_user should be unconditionally ignored as part of the constructor.

as far as if this "ordering" is something being looked for in things like `dir(cls)` or `mapper.attrs`, I would not change any of that.

**Reply by CaselIT** [reply]
Re looking the example you are actually right. I initially applied dataclass on every class, but `_BaseMixin` is not one.

I'll open an issue

**Reply by zzzeek** [reply, marked as answer]
see #9350.   there's unfortunately no fix for this that will not break backwards compatibility.    i think we should start by warning for those fields that are incorrectly being applied.

**Reply by CaselIT** [reply]
I've opened https://github.com/sqlalchemy/sqlalchemy/issues/9351

It seems there is also an issue with annotations of the `__init__`

**Comment by wu-clan** [comment]
It can be turned off now #9350
